### DISCUSSION TITLE
About the population size after iterations in NSGA-II

### DISCUSSION THREAD

**Discussion by Kazusaisjustice** [anyoptimization/pymoo#291] [Q&A] [answered]
Hello,
I have used the pymoo to solve a multi-object problem but I found sometimes the final population size is not the same as what I set at beginning and I still don't know the reason. In addition, in my multi-object problem I only increased the number of variables but I just need a less iteration to get the same population size as the beginning, I also don't know why. Thanks for your reading.

**Comment by blankjul** [comment, marked as answer]
In NSGA-II, the population size is always kept the same. You probably refer to the result object, which filters the non-dominated solutions from the population. Let us say `pop_size=100` then `len(res.pop)==100` but `len(res.opt)<=100`will hold. 
If you run the algorithm long enough:  finally, almost all solutions will be non-dominated in the current population and it will become the same.

**Reply by Kazusaisjustice** [reply]
You mean if the iteration goes enough, the final population must be the same? But in the principle of NSGA-II, after non-dominated sorting and crowding distance sorting no matter how many iteration times go the final size should be the same as the original parents' size, isn't it? So why the final size is not 100 (while we set up  pop_size=100), if the iteration time isn't long enough?

**Reply by bandundu** [reply]
Can you maybe provide a code example as I am not really sure what you are referring to.

**Reply by Kazusaisjustice** [reply]
Actually, I just hope to know why we cannot get the same size as the original parents', if the iteration times are not enough. In my opinion, no matter how many iterations, the final population size should be the same, but may cannot converge.

**Reply by blankjul** [reply]
Can you post a MWE (Minimum Working Example) to make your point? In NSGA2, the population object `algorithm.pop` should always be of size `pop_size` (there are rare cases where this might not hold, for instance when the search space is very small and duplicate elimination can mess up things).

**Reply by Kazusaisjustice** [reply]
Oh, I see! I know why it happened. Thank you very much!
